File: src/LiveAlert.Core/AlertQueue.cs
using System.Linq;

namespace LiveAlert.Core;

public sealed class AlertQueue
{
    private readonly List<AlertQueueItem> _pending = new();

    public AlertQueueItem? DequeueNext()
    {
        if (_pending.Count == 0) return null;
        _pending.Sort(Compare);
        var item = _pending[0];
        _pending.RemoveAt(0);
        return item;
    }

    public void Enqueue(AlertQueueItem item)
    {
        if (_pending.Any(x => x.VideoId == item.VideoId)) return;
        _pending.Add(item);
    }

    public void RemoveByVideoId(string videoId)
    {
        _pending.RemoveAll(x => x.VideoId == videoId);
    }

    public bool Contains(string videoId)
    {
        return _pending.Any(x => x.VideoId == videoId);
    }

    private static int Compare(AlertQueueItem a, AlertQueueItem b)
    {
        var time = a.DetectedAt.CompareTo(b.DetectedAt);
        if (time != 0) return time;
        return a.Event.AlertIndex.CompareTo(b.Event.AlertIndex);
    }
}

public sealed record AlertQueueItem(AlertEvent Event, DateTimeOffset DetectedAt)
{
    public string VideoId => Event.VideoId;
    public AlertConfig Alert => Event.Alert;
}

File: src/LiveAlert.Electron/src/core/alert-queue.ts
import { AlertEvent } from './alert-monitor';
import { AlertConfig } from './config-models';

export interface AlertQueueItem {
  event: AlertEvent;
  detectedAt: number;
  videoId: string;
  alert: AlertConfig;
}

export class AlertQueue {
  private pending: AlertQueueItem[] = [];

  dequeueNext(): AlertQueueItem | null {
    if (this.pending.length === 0) return null;
    this.pending.sort((a, b) => {
      const time = a.detectedAt - b.detectedAt;
      if (time !== 0) return time;
      return a.event.alertIndex - b.event.alertIndex;
    });
    return this.pending.shift()!;
  }

  enqueue(event: AlertEvent): void {
    if (this.pending.some((x) => x.videoId === event.videoId)) return;
    this.pending.push({
      event,
      detectedAt: event.detectedAt,
      videoId: event.videoId,
      alert: event.alert,
    });
  }

  removeByVideoId(videoId: string): void {
    this.pending = this.pending.filter((x) => x.videoId !== videoId);
  }

  contains(videoId: string): boolean {
    return this.pending.some((x) => x.videoId === videoId);
  }
}
